adr/ADR-20.md

```diff
@@ -0,0 +1,73 @@
+
+Each JetStream only listens to default API subjects with the prefix `$JS.API`.
+Thus, when the client uses `$JS.API`, it communicates with the JetStream, local to it's account and domain.
+
+To avoid traffic going to some other JetStream the following mechanisms are in place:
+1. Account: Since the API has to be imported with an altered prefix, the request will not cross account boundaries. 
+    In a JetStream enabled account, an import without a prefix set will result in an error as JetStream is imported as well and we error on import overlaps. 
+2. Domain: On leaf node connections, the nats server adds in denies for `$JS.API.>`.
+
+When the client library sets an API prefix, all API subjects, the client publishes and subscribes to start with that instead of `$JS.API`.
+As a result the API traffic will not end up in the local JetStream as that alway only ever subscribes to `$JS.API....`
+
+As messages and subscriptions cross boundaries the following happens:
+1. Accounts: When crossing the import, the API prefix is stripped and replaced with `$JS.API`.
+2. Domain: When crossing the leaf node connection between domains, an automatically inserted mapping strips the API prefix and replaces it with `$JS.API`.
+
+This JetStream disambiguation mechanism needs to be added to materialized views such as KV and object store as well.
+Specifically, we need to tag along on the same API prefix. 
+Setting different values to reach the same JetStream is a non starter.
+
+## Design
+
+The first token of any API or materialized view is considered the default API prefix and will mean have local semantics.
+Thus, for the concrete views we treat the tokens `$KV` and `$OBJ` as the respective default API prefixes. 
+For publishes and subscribes the API just replaces the first token with the specified (non default) prefix. 
+
+To share API access across accounts this is sufficient as account export/import takes care of the rest. 
+To access an API in the same account but different domain, the `nats-server` maintaining a leaf node connection will add in the appropriate mappings from domain specific API to local API and deny local API traffic.
+
+## KV Example
+
+Assume the API prefix for JetStream has been set to `JS.from-acc1`.
+For JetStream specific API calls, the local API prefix `$JS.API` has to be replaced with `JS.from-acc1`.
+For the KV API we prefix `$KV` with `JS.from-acc1`, resulting in `JS.from-acc1.$KV`. 
+Thus, in order to put `key` in the bin `bin`, we send to `JS.from-acc1.$KV.bin.key` 
+
+When crossing the account boundaries, this is then translated back to `$KV.bin.key`.
+Thus, the underlying stream still needs to be created with a subscription to `$KV.bin.key`. 
+
+Domains are just a special case of API prefixes and will work the same way.
+The API prefix `$JS.<domain-name>.API` will lead to `$JS.<domain-name>.API.$KV.bin.key`.
+
+## Consequences 
+
+The proposed change is backwards compatible with materialized views already in use. 
```

Comment:
Generally not super apposed to this, but we have this problem with any stream already in all cases.  I dont really want to fix the problem just for our special streams.

Reply:
Yeah, since those ($KV,$OBJ) are just (parts of) stream names, why would we change the name. Why are these different than any other stream and how it is handled within / across a domain? The API calls we make just treat these streams like any other.

Reply:
These are for the stream listen subject not names - but yeah, that's the general sentiment.

Reply:
Which problem are you referring to?

Reply:
If 2 streams in 2 domains have the same interest subjects both get the message.  This is the problem you are addressing with the `$KV...` subjects.  But users have the same problem with their own streams.

Reply:
We are not changing any streams names.
We just change the subjects you use to interact with a resource such that in the current account or domain, they are not received. (assuming they are different)
On entering the target account or entering the target domain, we change them back to what they would have been, had the client been connected and sent them there directly.

What this is about is talking to the right JetStream. Nothing else.
And it is doing so in the very same way we already do.
What I suggest here is the adaptation of an existing technique.

That this also solves the stream matching issue is mere coincidence because the subjects used to subscribe are part of the API.
To me the problem of a message matching possibly multiple streams is a very different one.
One, that frankly only needs addressing for domains.

I do have ideas on how to solve this, but I don't think this ADR is the forum for that.

Reply:
Yes, please read again, I did mention subjects not stream names. 

Since this is a generic problem for any stream Derek said we should instead find a generic solution.  We should not solve this only for our streams.

Reply:
Scott did mention stream names

Reply:
While I see this being an issue for domains, cross accounts I don't think there is anything we can do. It's a different account, so would you not store a message in account B, just because it was stored in account A already?
Since this issue does not really apply across accounts, we should support the api prefix, just because it is the current solution. 
What is your opinion on that?

Reply:
Thus, supporting the prefix has to happen just to support the cross account use case.

Reply:
As for the cross domain use case, in a config based setup, after the push consumer change when cross account stuff is simplified, a user could just have a transit account, from which it only exports API traffic into the actual account used. Isolated like this, all the traffic is fine.

In a jwt based world we'd need the templating mechanism, such that the domain name can be adapted to the domain name of the servers the account is running in. 

If we wanted to push this even further and make it less export/import intensive, I would frankly add a flag to the account named local_scope. When true, all traffic in and out of a leaf node connection where the neighboring domain names differ is denied, except for traffic sent to the domain specific api prefix. 
An application can then choose the setting according to the behavior it requires.
Frankly, this has nothing to do with JetStream either. 
If I have a service foo and set local_scope true, I can reach the service local to main domain under foo and the one offered under the same name in a different domain under say: $EXT.domain.foo

With this, the user can then write domain scoped apps and communicate that to us with a setting. 
We take over the addressing portion externally from their app, with a name that already exists.

Reply:
I bet this made you cringe. 
I also would not want to have the resulting hour long discussions about that in the critical path for KV or object. 

Therefore, I would focus on this for now:
We need to support the API prefix in order to deal with the cross account use case.
So the necessary client changes have to happen just because of that.

Domains are a special case of the cross account use case that require a two line server change.

Since one side has to happen anyway and the other side is small and straight forward, I would try to not open a whole other can of worms just yet.

Reply:
For cross account, we have three main patterns for base JetStream, and hence upper level abstractions like KV. 

1. Placing things into a stream
2. Consuming from a stream via pull
3. Consuming from a stream via push (which is under a design change atm)

All 3 above are (or in case of 3 will be) a service export, meaning the importing account already has control of subject collisions since during import you can remap these.

So, in case of KV, let's say account A has a bucket called foo. So that means the subject is `$KV.foo.>`.
Let's say account B also has a foo KV bucket but wants to access account A. 

During import I could say remap the account A service (not stream because we get a PubAck response), I could ask it to show up as `$KV.foo_A.>`. 

Now I am not sure if we could, inside account B, simple say bind to `foo_A`, but that should work imo.

What I would like to understand better is the following, is the suggested change part of simply making binding to an asset properly leveraging the import remapping we already have, or is this something different altogether?

Reply:
@derekcollison It is leveraging what we already have and is following the pattern already established form streams.